### PR TITLE
feat: upgrade to Angular 17

### PR DESCRIPTION
**Pull request by Tommy228**
- ran ng update @angular/cli @angular/core
- updated eslint dependencies
- updated jest dependencies

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/spectator/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #631 

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

## Other information

Angular recently introduced [TestBed.flushEffects()](https://angular.io/guide/signals#effects) which is a helper function to execute pending [effects ](https://angular.io/guide/signals#effects). I think it makes sense to add `spectator.flushEffects()` but the method is still under dev preview, so maybe not the right time?

### CODE REVIEW

**Comment by chimurai** [comment]
`peerDependencies` needs to be updated:

https://github.com/ngneat/spectator/blob/9478289fd35894f9e9fda2785c459d2bfdb276e0/projects/spectator/package.json#L32-L34

**Review by chimurai — CHANGES_REQUESTED** [review]
`peerDependencies` needs to be updated:

https://github.com/ngneat/spectator/blob/9478289fd35894f9e9fda2785c459d2bfdb276e0/projects/spectator/package.json#L32-L34

**Comment by NetanelBasal** [comment]
You can add spectator.flushEffects() and also introduce a an API to test defer views.

**Comment by chimurai** [comment]
Running `yarn build` fails:

```shell
$ yarn build
yarn run v1.22.21
$ ng build --configuration production && yarn build:schematics && yarn copy:schematics && yarn copy:docs
Building Angular Package

------------------------------------------------------------------------------
Building entry point '@ngneat/spectator/internals'
------------------------------------------------------------------------------
✖ Compiling with Angular sources in Ivy partial compilation mode.
error TS2688: Cannot find type definition file for 'jasmine'.
  The file is in the program because:
    Entry point of type library 'jasmine' specified in compilerOptions
error TS2688: Cannot find type definition file for 'node'.
  The file is in the program because:
    Entry point of type library 'node' specified in compilerOptions

error Command failed with exit code 1.
```


Updating `jest` and `@types/jest` to latest and patching `spectator/tsconfig.lib.json` seems to fix it:

`package.json`:
- `"@types/jest": "29.5.10",`
- `"jest": "29.7.0",`


```diff
index adeaf7c4..8566c2b4 100644
--- a/projects/spectator/tsconfig.lib.json
+++ b/projects/spectator/tsconfig.lib.json
@@ -6,10 +6,10 @@
     "declaration": true,
     "inlineSources": true,
     "typeRoots": [
+      "../../node_modules/@types",
       "./typings"
     ],
     "types": [
-      "jasmine",
       "jest",
       "node"
     ],
```

**Comment by Tommy228** [comment]
Thanks, it's fixed now. I've tried running this on a big project under ng17 with `yarn link` but I'm having some typing errors
```
Property 'service' does not exist on type 'SpectatorService<TestService>'.
```
The generated `dist` looks fine though. I'm trying to understand why, maybe it's just an issue with my local setup.

**Comment by NetanelBasal** [comment]
@Tommy228 try with [verdaccio](https://verdaccio.org/)

**Comment by Tommy228** [comment]
Thanks, that worked. I didn't know about it, pretty handy :) 

I have just tried the upgrade on an Angular 17 project with 4000+ tests with Spectator and Jest, works fine.

File: projects/spectator/tsconfig.lib.json [hunk 1]
```diff
@@ -6,10 +6,10 @@
     "declaration": true,
     "inlineSources": true,
     "typeRoots": [
+      "../../node_modules/@types",
```
Comment (NetanelBasal) [inline projects/spectator/tsconfig.lib.json:9]: Can you explain this change, please?

Comment (chimurai) [reply projects/spectator/tsconfig.lib.json:9]: typescript got upgraded from `5.0.4` to `5.2.2`

Typescript 5.1 introduced a change in `typeRoots` behaviour:
> Explicit `typeRoots` Disables Upward Walks for `node_modules/@types`

More info: https://devblogs.microsoft.com/typescript/announcing-typescript-5-1/

Comment (chimurai) [reply projects/spectator/tsconfig.lib.json:9]: Looking at the Microsoft documentation the order of `typeRoots` should be corrected:

```
    "typeRoots": [
      "./typings",
      "../../node_modules/@types"
    ],
```

File: projects/spectator/tsconfig.lib.json [hunk 2]
```diff
@@ -6,10 +6,10 @@
...
       "./typings"
     ],
     "types": [
-      "jasmine",
```
Comment (NetanelBasal) [inline projects/spectator/tsconfig.lib.json:12]: Why we remove jasmine from here?

Comment (chimurai) [inline projects/spectator/tsconfig.lib.json:12]: `jasmine` type should remain.

**Review by chimurai — CHANGES_REQUESTED** [review]

**Comment by Tommy228** [comment]
My project uses jest, so I didn’t test it under jasmine. Indeed, we should add it back. I can’t update the PR before a few days so feel free to do it if you want to.

**Comment by lneumeier** [comment]
> My project uses jest, so I didn’t test it under jasmine. Indeed, we should add it back. I can’t update the PR before a few days so feel free to do it if you want to.

I created a new pull-request.
https://github.com/ngneat/spectator/pull/633

**Pull request closed without merging**